toolkit/offchain/src/d_param.rs

```diff
@@ -0,0 +1,234 @@
+		let collateral = make_utxo(7u8, 0, 72345678);
+		let payment_utxo = make_utxo(4u8, 1, 4000000);
+		let payment_addr =
+			Address::from_bech32("addr_test1vqp4a7r0zc3pw2qkhw0fz6h2s6grktydxtrj3t2unw2890sfgt0kq")
+				.unwrap();
+		let validator_addr =
+			Address::from_bech32("addr_test1wpha4546lvfcau5jsrwpht9h6350m3au86fev6nwmuqz9gqer2ung")
+				.unwrap();
+		let pub_key_hash = hex!("035ef86f1622172816bb9e916aea86903b2c8d32c728ad5c9b9472be").into();
+		let ex_units = ExUnits::new(&10000u32.into(), &200u32.into());
+
+		let tx = mint_token_tx(
+			&PlutusScript::new_v2(hex!("4d4c01000022223212001375a009").to_vec()),
+			&DParameter { num_registered_candidates: 30, num_permissioned_candidates: 40 },
+			&pub_key_hash,
+			&[collateral],
+			&[payment_utxo],
+			NetworkIdKind::Testnet,
+			&protocol_parameters(),
+			ex_units.clone(),
+		)
+		.unwrap();
+
+		let body = tx.body();
+		let inputs = body.inputs();
+		// Payment input goes to inputs
+		assert_eq!(
+			inputs.get(0).to_string(),
+			"0404040404040404040404040404040404040404040404040404040404040404#1"
+		);
+		// Collateral input goes to collaterals
+		assert_eq!(
+			body.collateral().unwrap().get(0).to_string(),
+			"0707070707070707070707070707070707070707070707070707070707070707#0"
+		);
+		let outputs = body.outputs();
+		// There is a change for payment
+		assert!(outputs.into_iter().find(|o| o.address() == payment_addr).is_some());
+		// There is 1 d-param token in the validator address output
+		let script_output = outputs.into_iter().find(|o| o.address() == validator_addr).unwrap();
+		let token_policy_id =
+			ScriptHash::from(hex!("6fdad2bafb138ef29280dc1bacb7d468fdc7bc3e93966a6edf0022a0"));
+		assert_eq!(
+			script_output
+				.amount()
+				.multiasset()
+				.unwrap()
+				.get_asset(&token_policy_id, &empty_asset_name(),),
+			1u64.into()
+		);
+		let expected_plutus_data = {
+			let mut list = PlutusList::new();
+			list.add(&PlutusData::new_integer(&40.into()));
+			list.add(&PlutusData::new_integer(&30.into()));
+			PlutusData::new_list(&list)
+		};
+		assert_eq!(script_output.plutus_data().unwrap(), expected_plutus_data);
+		// This token is minted in the transaction
+		let mint = tx.body().mint().unwrap();
+		assert_eq!(
+			mint.get(&token_policy_id)
+				.unwrap()
+				.get(0)
+				.unwrap()
+				.get(&empty_asset_name())
+				.unwrap(),
+			Int::new_i32(1)
+		);
+		// Redeemer is set
+		let redeemers = tx.witness_set().redeemers().unwrap();
+		assert_eq!(redeemers.len(), 1);
+		let redeemer = redeemers.get(0);
+		assert_eq!(redeemer.tag(), RedeemerTag::new_mint());
+		assert_eq!(redeemer.index(), 0u64.into());
+		assert_eq!(redeemer.data(), PlutusData::new_empty_constr_plutus_data(&0u64.into()));
+		assert_eq!(redeemer.ex_units(), ex_units);
+		// script_data_hash check
+		assert_eq!(
+			tx.body().script_data_hash().unwrap().to_hex(),
+			"5b95e874a40a87b017ee7827a7dccf7331d2b647190eddcde7f0edaba4393662"
+		);
```

Comment:
I am now working on https://input-output.atlassian.net/browse/ETCM-8837 . Any conclusions to that ticket will be applied in this test